### DISCUSSION TITLE
To use middleware you must provide a hostname and port to the Next.js Server

### DISCUSSION THREAD

**Discussion by LuckeeDev** [vercel/next.js#33835] [Help] [answered]
I get this error whenever I run my app: `To use middleware you must provide a hostname and port to the Next.js Server`. How can I fix this? Where should I provide `hostname` and `port`?

For context, Next is on `v12.0.8`.

**Comment by wwwdepot** [comment, marked as answer]
See Custom Server docs https://nextjs.org/docs/advanced-features/custom-server

```js
const hostname = 'localhost'
const port = 3000
// when using middleware `hostname` and `port` must be provided below
const app = next({ dev, hostname, port })
```

**Reply by LuckeeDev** [reply]
The issue is that I'm not using a custom server. Or does NX count as a custom server?

**Reply by wwwdepot** [reply]
Just tested a new NX + Next.js project, there's no warning/error when using middleware. Please provide a reproduction repo.

**Reply by LuckeeDev** [reply]
I have just been able to fix the issue by providing a `hostname` to my NX config. Thanks for helping me out!

**Reply by prabhatthapa** [reply]
But why do we need this change?

**Reply by drkstr101** [reply]
I'm getting this error whenever `experimental.concurrentFeatures` is true. I've tried setting as both as env variable and in next config to no avail.

https://github.com/drkstr101/waweb/pull/6 (gitpod enabled)

**Reply by jere-co** [reply]
> I have just been able to fix the issue by providing a `hostname` to my NX config. Thanks for helping me out!

@LuckeeDev How did you provide them exactly? Dealing with the same situation here.

**Reply by LuckeeDev** [reply]
@jere-co in your `project.json`, make sure you have this

```json
"targets": {
  "serve": {
    "executor": "@nrwl/next:server",
    "options": {
      "buildTarget": "projectName:build",
      "dev": true,
      "port": 3000,
      "hostname": "localhost"
    }
  }
}
```

**Reply by drewrawitz** [reply]
Thanks, @LuckeeDev!  What about when it comes time to deploy your nx app?  I think in your `production` configuration, you'd also have to provide a `hostname` and `port`, but I can't hardcode those values. I need them to come from my `.env` file 🤔

**Reply by LuckeeDev** [reply]
I'm not sure about it, but I think you won't need to provide a hostname since you'll be running Next with the `next start` command in production and not from a custom server (like NX does). You should probably try and see what works

**Comment by nickyoung87** [comment]
I haven't looked into this yet deeply to find a solution, but I wanted to note here that I have ran into the same issue when trying to use the middleware on firebase hosting. I saw someone said it should be fine in production, but this seems to go against that.

I will post back here if I find any solutions for it, but if anyone else has info then would be glad to hear that too.

**Reply by nickyoung87** [reply]
I haven't found anything conclusive yet to get this working. I will note that it definitely doesn't work straight out of the box. I will keep searching and see if I can find a workaround. Would be really useful to use the middleware feature with Firebase.

**Comment by Jingeun** [comment]
If It doesn't work in docker environment, put 0.0.0.0 in the hostname.
Then it works out!
